File: 0023-Merge-k-Sorted-Lists/0023.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        import heapq 
        result = ListNode(-1)
        cur = result
        p = list()
        x = 0
        for i in lists:
            while i:
                heapq.heappush(p, (i.val, x, i))
                i = i.next
                x += 1

        while p:
            cur.next = heapq.heappop(p)[1]
            cur = cur.next

        return result.next

def createList():
    head = ListNode(0)
    cur = head
    cur.next = ListNode(0)
    cur.next.next = ListNode(1)
    cur.next.next.next = ListNode(2)
    return head
            
def printList(head):
    cur = head
    while cur != None:
        print(cur.val, '-->', end='')
        cur = cur.next

    print('NULL')

if __name__ == "__main__":
    l1 = createList()
    l2 = createList()
    l3 = createList()
    printList(l1)
    l = list(l1)
    l.append(l2)
    l.append(l3)
    res = Solution().mergeKLists(l)
    printList(res)

File: src/0023-Merge-k-Sorted-Lists/0023.cpp
#include <string>
#include <iostream>
#include <vector>
#include <functional>
#include <queue>
#include <deque>
using namespace std;

//  Definition for singly-linked list.
struct ListNode 
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution 
{
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) 
    {
        priority_queue<pair<int, ListNode*>, deque<pair<int, ListNode*>>, greater<pair<int, ListNode*>>> min_heap;
        for (ListNode* l : lists) 
        {
            if (l) min_heap.push(make_pair(l->val, l));
        }
        ListNode *head = new ListNode(-1);
        ListNode *cur = head;
        while (!min_heap.empty()) 
        {
            cur->next = min_heap.top().second;
            cur = cur->next;
            min_heap.pop();
            if (cur->next) min_heap.push(make_pair(cur->next->val, cur->next));
        }
        ListNode *ret = head->next;
        delete head;
        return ret;
    }
};

int main()
{
    return 0;
}